homeassistant/components/tuya/__init__.py

```diff
@@ -192,4 +149,21 @@ def async_remove_device(self, device_id: str) -> None:
         )
         if device_entry is not None:
             device_registry.async_remove_device(device_entry.id)
-            self.device_ids.discard(device_id)
+
+
+class TokenListener(SharingTokenListener):
+    """Token listener for upstream token updates."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        entry: ConfigEntry,
+    ) -> None:
+        """Init TokenListener."""
+        self.hass = hass
+        self.entry = entry
+
+    def update_token(self, token_info: str) -> None:
+        """Update token info in config entry."""
+        data = {**self.entry.data, "token_info": token_info}
+        self.hass.config_entries.async_update_entry(self.entry, data=data)
```

Comment:
Will this run in async or sync context?

Reply:
sync, wrapped it and raised a job from here